for-loops.md

```diff
@@ -0,0 +1,203 @@
+  `for` loops as shown by the grammar above.
+
+- Don't support "down" loops. This would mean `downto` does not need to be reserved.
+
+- Use `..` instead of `to` for "up" loops. Without the invention of new syntax
+  for intervals written in the "opposite order", this excludes "down" loops.
+  If `..` is used in the syntax, then it would also seem natural to write `in`
+  instead of `:=`, like this:
+
+  ```
+  for i in lo .. hi
+  ```
+
+# Prior art
+[prior-art]: #prior-art
+
+Most popular languages have some form of `for` loop. In C, the first line of the
+proposed `for` loops would look like:
+
+``` C
+for (int i = lo; i < hi; i++)
+```
+
+ignoring issues of types, issues of scopes, the issue of re-evaluating `hi`, and
+the issue of prohibiting other updates of `i`.
+
+"Down" loops written in C look like:
+
+``` C
+for (int i = hi; lo < --i; )
+```
+
+The Eiffel programming language supports loop invariants. It has an `until`
+condition, which says when to terminate, rather than the negated termination
+condition of standard `while` loops and C's `for` loop. The update of the
+loop index is provided in the program text, as for common `while` loops.
+
+Why3 supports the OCaml-like `for` loop
+
+```
+for i = lo to hi do invariant { J } Body done
+```
+
+It also has a `downto` variant. In this form, the upper bound `hi` is still
+inclusive; that is, the automatically generated loop invariant is
+
+```
+lo <= i <= hi + 1
+```
+
+Another unique and verification-friendly feature of Why3's `for` loops
+is that if the values of `lo` and `hi` are such that the loop body will not
+iterate at all, then the entire loop is skipped, including the checking of
+the invariant on entry to the loop.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Are "down" loops important to include in the language?
+- Is it okay to exclude loops there the loop index changes by something other than `+1` or `-1`?
+- Is the `..` token to be preferred over the `to` keyword?
+- If `..` is used and "down" loops are important, should there be an additional keyword like `reverse`
+  added to the syntax somewhere?
+- `while` loops are nice in teaching. They make it clear where the loop index is declared and what
+  values it may have. You can also see the update of the loop index, which is useful if you want
+  to use weakest preconditions to compute the needed values of variables near the end of the loop
+  body. Would the availability of `for` loops lead teachers and students astray by luring them
+  into using `for` loops before loop invariants are clearly understood?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+Another kind of loop that is missing from Dafny are `foreach` loops, which loop over all
+values of some collection or `iterator`. If such were ever introduced, a syntax like
+
+```
+foreach x in S
+  invariant J
+{
+  Body
+}
+```
+
+is still available.
```

Comment:
FYI, Why3 has a `for each` loop with syntax
```
for <pat> in <expr> with <Module> do invariant {...} variant {...} <body> done
```
It roughly desugars to
```
let it = <Module>.create <expr> in
try while true do invariant {...} variant {...} let <pat> = <Module>.next it in <body> done
with <Module>.Done -> () end
```
which means that <Module> is providing
* a function `create` (to create the iterator from <expr>)
* a function `next` (to get the next value)
* an exception `Done` (to signal termination of the iteration)

Reply:
Regarding the way to specify `create` and `next`, you might be interested in this NFM 2016 paper:
https://hal.inria.fr/hal-01281759v2/document

Reply:
What is the reason not to use just `foreach` loops (using the `for` keyword seems fine)?

`[lo .. hi]` in `for i in [lo .. hi]` is already in Dafny. That'd allow for arbitrary stepping using sequence comprehensions such as in `for x in seq(5, i => i*i)`.

The versions with just `-1` and `+1` updates seem restrictive to me.